pkg/controlplane/control/manager.go

```diff
@@ -193,10 +233,87 @@ func (m *Manager) AddImport(ctx context.Context, imp *v1alpha1.Import) (err erro
 	return m.addImportService(ctx, imp, userService)
 }
 
+// deleteClDataplane deletes pod to ipToPod list.
+func (m *Manager) deleteEndpoint(esID types.NamespacedName) {
+	if strings.Contains(esID.Name, dpapp.Name) {
+		m.logger.Warnf("Detected cl-dataplane(%s) endpointslice delete in namespace: %s.", esID.Name, esID.Namespace)
+	}
+}
+
+// addPod adds or updates pod to ipToPod and podList.
+func (m *Manager) addEndpoint(ctx context.Context, eslice *discv1.EndpointSlice) {
+	m.podLock.Lock()
+	defer m.podLock.Unlock()
+
+	if eslice.Labels[labelServiceName] != dpapp.Name {
+		return
+	}
+	m.logger.Infof("Detected cl-dataplane(%s) endpointslice add/update in namespace: %s", eslice.Name, eslice.Namespace)
+	m.endpoints[eslice.Namespace] = eslice.Endpoints
+	// Retrieve endpointslices managed by clusterlink in the namespace and update them
+	labelSelector := labels.Set(map[string]string{
+		labelESManagedBy: appName,
+	})
+	listOpts := []client.ListOption{
+		client.InNamespace(eslice.Namespace),
+		client.MatchingLabels(labelSelector),
+	}
+
+	endpointSliceList := &discv1.EndpointSliceList{}
+	err := m.client.List(ctx, endpointSliceList, listOpts...)
+	if err != nil {
+		m.logger.Errorf("Failed to list endpointslices managed by %s", appName)
+		return
+	}
+
+	for i := range endpointSliceList.Items {
+		es := endpointSliceList.Items[i]
+		m.logger.Infof("Updating endpointslice %s", es.Name)
+		es.Endpoints = eslice.Endpoints
+		err := m.client.Update(ctx, &es)
+		if err != nil {
+			m.logger.Errorf("Failed to update endpointslice %s: %v.", appName, err)
+		}
+	}
+}
```

Comment:
I think this entire section should be replaced with the following logic:
1. List all "merge=true" imports.
2. For each such import, make sure the import endpointslices are consistent with the current dataplane endpoints.

Your current code assumes the dataplane endpointslices are fixed (same number of endpoint slices, and names cannot change).
I don't think this assumption is good. Specifically, for example, client can scale up/down the number of dataplane replicas.

I would implement step 1 by either maintaing an in-memory map of all merge-type imports, or maybe even better by moving out the merge field (which is currently in the importspec) to be a label. That way you can use a label selector to list just the merge-type imports.

Step 2 would be simply a call to `addImportEndpointSlice`.

Reply:
That's the same logic which is implemented by the section.
1) First we retrieve the endpoint slices using `managed-by` label , which is set when we do `merge=true`
2) The subsequent for loop iterates of the `endpointsliceList` to make sure its consistent.

I don't understand the assumptions. What do you mean by dataplane endpoint slices are fixed? We are only tracking cl-dataplane's endpointslice. We do not track the name of the endpointslice, its tracked using `service-name=cl-dataplane` label , which is the right way.